### ISSUE TITLE
Fire Emblem Three Houses: Pink/Red Artifacting on Certain Blue/Purple Models

### ISSUE THREAD

**Issue by LinkH334** [yuzu-emu/yuzu#7653]
Hi everyone, this is my first time reporting a bug for Yuzu, so I hope I'm doing this correctly. If I've done anything incorrectly or this isn't the correct place for this kind of report, please let me know! Anyway, on to the actual bug report.

While playing Fire Emblem: Three Houses, I've encountered this pink/red artifacting on certain models shaded/textured certain shades of blue and purple. Kind of hard to describe, so just check the screenshots I've attached. This happened no matter what settings I changed (in both OpenGL and Vulkan, native resolution or upscaled, GPU Accuracy set to Normal, High, or Extreme, Multi-core CPU Emulation On or Off, etc.). I do use a few mods with Three Houses (the 60 FPS mod and the Remove Outlines mod), but the issue persists when they are disabled.

My system specs:
OS: Windows 10 Home Version 21H2 (updated just today, then retested the issue)
CPU: 11th Gen Intel(R) Core(TM) i5-11600K @3.90GHz
GPU: NVIDIA RTX 3080ti Founder's Edition
RAM: 32GB
 
This issue was observed on the most recent mainline build of Yuzu (version 864). It may also be worth noting that I experience the same issue in Ryujinx, as in visually identical (also the most recent build there).

If there's any additional info or files you need me to provide, please let me know and I'll get back to you as soon as I can. If you'd like me to provide a save file right in front of one of the characters that exhibit this problem so it can be easily reproduced without having to play parts of the game, I can do that, too!

Log: [yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/7795111/yuzu_log.txt)

 Screenshots:

![010055d009f78000_2021-12-29_21-56-07-006](https://user-images.githubusercontent.com/96894850/147798554-33751aed-6616-4857-b84b-1dd687feb49a.png)
Most noticeable in Felix's hair.

![010055d009f78000_2021-12-29_21-54-58-136](https://user-images.githubusercontent.com/96894850/147798560-0eb85b1d-76b1-4c19-9215-b5193c1eab24.png)
Also visible in Shamir's hair.

![010055d009f78000_2021-12-29_22-02-37-299-min](https://user-images.githubusercontent.com/96894850/147798817-2d6f365d-c5b0-4070-a238-561dcce117e9.png)
And Lorenz's.

![010055d009f78000_2021-12-30_07-53-05-533-min](https://user-images.githubusercontent.com/96894850/147798819-decc2bf0-e77f-4c17-972e-5eb99761ad32.png)
Faintly visible on the blue part of Seteth's outfit.

![010055d009f78000_2021-12-29_21-55-25-902-min](https://user-images.githubusercontent.com/96894850/147798821-a7fd8406-64c7-47d3-bfc8-c0a806b8bb03.png)
And finally, also faintly visible on the scabbard of the dagger on Byleth's outfit.

**Comment by Bjorn29512** [comment]
Addendum : This issue is Ampere (30 Series) exclusive

**Comment by H-Toine** [comment]
I've had this happen to me for a while too, running a GTX 1080. Still happens in yuzu 1166
![yuzu_eHIDVbDsvS](https://user-images.githubusercontent.com/25739208/190412228-15782181-80d9-4714-b79b-cdccd0d074d4.png)

**Comment by Bjorn29512** [comment]
@LinkH334  mind sharing a save where it can be conveniently reproduced?

**Comment by LinkH334** [comment]
@Bjorn29512 Yeah, no problem. I've zipped and attached all my save files, since it doesn't seem like an individual save loads properly on its own. Seems like it needs the system save file as well.

Regardless, you should be able to reproduce the issue easily on File 16. Just load into the Monastery and Felix should be right in front of you. Then just look at his hair in first person. The save is post-time skip. If you care about spoilers, I believe the auto-save is pre-time skip and you should be able to explore the Monastery and find Felix there.

I've just done the above myself on yuzu 1193 and can confirm the issue is still present for me. My hardware is the same as when I authored the post. If you like me to share a screenshot or video, or there's anything else you need from me, please let me know.


[FE3H Saves.zip](https://github.com/yuzu-emu/yuzu/files/9742302/FE3H.Saves.zip)

**Comment by Bjorn29512** [comment]
> @Bjorn29512 Yeah, no problem. I've zipped and attached all my save files, since it doesn't seem like an individual save loads properly on its own. Seems like it needs the system save file as well.
> 
> Regardless, you should be able to reproduce the issue easily on File 17. Just load into the Monastery and Felix should be right in front of you. Then just look at his hair in first person. The save is post-time skip. If you care about spoilers, I believe the auto-save is pre-time skip and you should be able to explore the Monastery and find Felix there.
> 
> I've just done the above myself on yuzu 1193 and can confirm the issue is still present for me. My hardware is the same as when I authored the post. If you like me to share a screenshot or video, or there's anything else you need from me, please let me know.
> 
> Edit because I realized I listed the wrong save file.
> 
> [FE3H Saves.zip](https://github.com/yuzu-emu/yuzu/files/9742302/FE3H.Saves.zip)

![image](https://user-images.githubusercontent.com/110204265/194941729-76dca2b8-8894-42d7-8022-3e8bf9bae393.png)

i am not noticing anything from my testing here are you using any anisotropic filtering ?

**Comment by Bjorn29512** [comment]
actually nvm i figured it out, only happens on open gl, i recommend switching to vulkan 
![image](https://user-images.githubusercontent.com/110204265/194943391-94387949-6288-4dde-acc4-947e7c2a6cc5.png)

**Comment by LinkH334** [comment]
Actually, I've primarily used Vulkan for this game since setting Yuzu up. I tested running FE3H in both when setting it up and Vulkan ran better. Both had this thread's issue. Just out of curiosity, I booted it up in OpenGL and the issue is still present in both backends for me.
![FE3H Yuzu OpenGL](https://user-images.githubusercontent.com/96894850/194989766-df3c73c7-b113-4369-86fe-1fcc33b3e36c.png)

Here's a screenshot in Vulkan, too.
![FE3H Yuzu Vulkan](https://user-images.githubusercontent.com/96894850/194989855-b24acee4-9c29-4c52-810d-7034a35b4743.png)

I also tried setting Anisotropic Filtering to "Default" (it was set to 16x) and no change, issue still present.

**Comment by Bjorn29512** [comment]
> I've had this happen to me for a while too, running a GTX 1080. Still happens in yuzu 1166 ![yuzu_eHIDVbDsvS](https://user-images.githubusercontent.com/25739208/190412228-15782181-80d9-4714-b79b-cdccd0d074d4.png) (Shamir's hair and if you look closely, Byleth's scabbard as well)
What about you try it on vulkan too with driver 512.95

**Comment by goldenx86** [comment]
It's caused by FP16 precision.
RDNA2 is unaffected under normal conditions, but if you force the card to instead use FP32 like it would on a Pascal or Polaris card, the issue shows up.

**Comment by Digitaln3ko** [comment]
Just to chime in, issue still exists, both on OpenGL and Vulkan on a 1070.

**Comment by Kevarhk** [comment]
I have a 3080 and the issue still persists on Yuzu 3400. I'm assuming there is no way to fix it?

**Comment by goldenx86** [comment]
No, NVIDIA needs to fix the precision of FP16 on Ampere and newer hardware.
This will only work correctly on Turing and RDNA2 GPUs for now.

**Comment by ThisNekoGuy** [comment]
This issue also happens with GLASM on Nvidia (on [Arch] Linux)

**Comment by SgabelloBello** [comment]
> No, NVIDIA needs to fix the precision of FP16 on Ampere and newer hardware. This will only work correctly on Turing and RDNA2 GPUs for now.

I have a 4090 and I have the same problem too, so that's not the core issue.

**Comment by louislam125** [comment]
dang, it's been 2 years and haven't been solved

**Comment by goldenx86** [comment]
NVIDIA continues to emulate FP16 on Ampere and newer.
Get an AMD GPU and it's solved.

**Comment by ThisNekoGuy** [comment]
> This issue also happens with GLASM on Turing (on [Arch] Linux)

@goldenx86 It happens ob my 2080 Ti, so it definitely happens on Turing too

**Comment by goldenx86** [comment]
Noted. That hints to a driver change/regression. Will investigate with a 1650.

**Comment by goldenx86** [comment]
@ThisNekoGuy can confirm, it's easy to reproduce.
Sadly the reason is the same I mentioned, we don't have control over what NVIDIA changes in their driver's code. Use Vulkan from now on.

**Comment by ThisNekoGuy** [comment]
> @ThisNekoGuy can confirm, it's easy to reproduce. Sadly the reason is the same I mentioned, we don't have control over what NVIDIA changes in their driver's code. Use Vulkan from now on.

- https://github.com/yuzu-emu/yuzu/issues/10766
- https://github.com/yuzu-emu/yuzu/issues/10690
- https://github.com/yuzu-emu/yuzu/issues/10660

Nvidia + Vulkan isn't a valid stable option for this game

**Comment by Squall-Leonhart** [comment]
> NVIDIA continues to emulate FP16 on Ampere and newer.

Nvidia isn't emulating FP16, its performed in the Tensor FMA where hardware intrinsics may try to optimise the shader expression, this might not always work properly which is why the "precise" qualifier available in GLSL will force the shader to be used exactly as written.
